### PR TITLE
Add (preliminary) support for dicts structure

### PR DESCRIPTION
**Pull request by ccamel**
This PR introduces preliminary support for **Dicts** in prolog, a named structure representing key-value pairs. The implementation largely follows the specifications outlined in [SWI-Prolog's documentation](https://www.swi-prolog.org/pldoc/man?section=bidicts).

## Feature overview

Here's some examples illustrating the feature.

```prolog
?- A = point{x:1,y:2}.x.
A = 1

?- S = segment{to:point{y:20, x:10}, from:point{y:2, x:1}}.to.x.
S = 10

?- A = point{x:1,y:2}.get(x).
A = 1

?- Tag = point, X = Tag{y:2, x:6}.
Tag = point
X = point{x:6,y:2}

?- [point{x: A}, point{x: B}] = [point{x:1}, point{x:2}].
A = 1
B = 2

?- S = segment{to:point{y:20, x:10}, from:point{y:2, x:1}}.get(to/x).
S = 10

?- A = point{x:1, y:2}.put(_{x:3}).
A = point{x:3,y:2}

?- A = point{b:2, d:4, f:6}.put([a:1, c=3, e(5)]).
A = point{a:1,b:2,c:3,d:4,e:5,f:6}
```

### Data structure

Dicts are currently represented as a **compound term** (`Compound`) using the functor `dict`. 
- The first argument is the **Tag**.
- Subsequent arguments form an array of **key-value pairs**, sorted by key.
- Keys are _atoms_, and values are Prolog _terms_.

**Rationale:**  
This representation is simple, easy to implement, and benefits from Prolog's pattern matching and unification mechanics due to its compatibility with compound terms.

#### Dict

The parser is adjusted to support reading dictionaries. The representation format mirrors SWI-Prolog:

```text
Tag{Key1:Value1, Key2:Value2, ...}
```

#### Functional notation for Dicts

Dictionaries can be accessed using functional notation (infix). For example, to retrieve the value associated with the key Key in the dictionary Dict, you can write Dict.Key.

To enable this, in case there is no current op-declaration for an operator `.` and the dot does not serve as operand, the parser interprets the dot as an infix operator that maps to `$dot/2`. This operator acts as _syntactic sugar_ for invoking the `op/3` function to evaluate dictionary operations during the compilation of clauses.

#### Compilation

During clause compilation, a desugaring step translates terms containing `$dot/2` into equivalent sub-goals with `./3`.

Example:

```prolog
?- S = segment{from:point{x:1, y:2}, to:point{x:10, y:20}}.from.x.
S = 1.
```

is translated into:

```prolog
?- .(segment{from:point{x:1, y:2}, to:point{x:10, y:20}}, from, _V0),
   .(_V0, x, _V1),
   =(S, _V1).
S = 1.
```

with the generated bytecode:

```text
get_var(0)
get_var(1)
get_var(2)
enter()
put_dict(5)
put_const(segment)
put_const(from)
put_dict(5)
put_const(point)
put_const(x)
put_const(1)
put_const(y)
put_const(2)
pop()
put_const(to)
put_dict(5)
put_const(point)
put_const(x)
put_const(10)
put_const(y)
put_const(20)
pop()
pop()
put_const(from)
put_var(0)
call(. /3)
put_var(0)
put_const(x)
put_var(1)
call(. /3)
put_var(2)
put_var(1)
call(= /2)
exit()
```

## Implemented features

This PR implements the following features for Dicts:
- Key-based value access using single or nested keys.
- Access via the `get(?KeyPath)` function (`t{a:x}.get(a)`, `t{a:t{b:x}}.get(a/b)`).
- Value modification using the `put(+New)` function (`point{x:1, y:2}.put(_{x:3})`, `point{x:1, y:2}.put([x=3])`).

See unit tests for a complete overview.

## Additional considerations

- Dicts are currently implemented as compound terms, but this representation might evolve for performance reasons.
- The parser modifications may cause compatibility issues with existing programs. While extensive unit tests cover a wide range of cases, edge cases may still arise.

## Future Work

Planned improvements include:

- [ ] Support for small integer keys (`S = point{6:1, x:12}`).
- [ ] Predefined functions:
  - [ ] `get(?KeyPath, +Default)`
  - [ ] `put(+KeyPath, +Value)`
- [ ] Support for user-defined functions.
- [ ] Predicates for managing dicts (see: [SWI-Prolog Dict Predicates](https://www.swi-prolog.org/pldoc/man?section=ext-dict-predicates)):
  - [ ] `is_dict(@Term)`
  - [ ] `is_dict(@Term, -Tag)`
  - [ ] `get_dict(+Key, +Dict, -Value, -NewDict, +NewValue)`
  - [ ] `dict_create(-Dict, +Tag, +Data)`
  - [ ] `dict_pairs(?Dict, ?Tag, ?Pairs)`
  - [ ] `dict_same_keys(?Dict1, ?Dict2)`
  - [ ] `put_dict(+New, +DictIn, -DictOut)`
  - [ ] `put_dict(+Key, +DictIn, +Value, -DictOut)`
  - [ ] `del_dict(+Key, +DictIn, ?Value, -DictOut)`
  - [ ] `+Select :< +From`
  - [ ] `select_dict(+Select, +From, -Rest)`
  - [ ] `+Dict1 >:< +Dict2`

### CODE REVIEW

File: engine/parser_test.go [hunk 1]
```diff
@@ -141,7 +141,7 @@ func TestParser_Term(t *testing.T) {
 		{input: `a + ().`, err: unexpectedTokenError{actual: Token{kind: tokenClose, val: ")"}}},
 		{input: `a * b + c.`, term: &compound{functor: atomPlus, args: []Term{&compound{functor: NewAtom("*"), args: []Term{NewAtom("a"), NewAtom("b")}}, NewAtom("c")}}},
 		{input: `a [] b.`, err: unexpectedTokenError{actual: Token{kind: tokenOpenList, val: "["}}},
-		{input: `a {} b.`, err: unexpectedTokenError{actual: Token{kind: tokenOpenCurly, val: "{"}}},
+		{input: `a {} b.`, err: unexpectedTokenError{actual: Token{kind: tokenLetterDigit, val: "b"}}},
```
[the remark this answers is not in the document]
Comment (ccamel) [reply engine/parser_test.go:144]: Oh yes, good idea! I focused on testing the interpreter, but you’re right. It's better to also target the _parser_. I’ll add tests for that.

However, be careful: your proposed test isn’t correct. Parser tests shouldn’t use the dictionary constructor but should instead directly validate the expected combination of terms produced by the parser. Here's what I'm going to add:

```go
{input: `tag{}.`, term: &dict{compound{functor: "dict", args: []Term{NewAtom("tag")}}}},
{input: `tag{k:v}.`, term: &dict{compound{functor: "dict", args: []Term{NewAtom("tag"), NewAtom("k"), NewAtom("v")}}}},
{input: `t.d.`,
	termLazy: func() Term {
		return &compound{functor: ",", args: []Term{&compound{functor: ".", args: []Term{NewAtom("t"), NewAtom("d"), lastVariable()}}, lastVariable()}}
	}},
{input: `X{}.`,
	termLazy: func() Term {
		return &dict{compound{functor: "dict", args: []Term{lastVariable()}}}
	},
	vars: func() []ParsedVariable {
		return []ParsedVariable{
			{Name: NewAtom("X"), Variable: lastVariable(), Count: 1},
		}
	},
},
{input: `t{k:V}.`,
	termLazy: func() Term {
		return &dict{compound{functor: "dict", args: []Term{NewAtom("t"), NewAtom("k"), lastVariable()}}}
	},
	vars: func() []ParsedVariable {
		return []ParsedVariable{
			{Name: NewAtom("V"), Variable: lastVariable(), Count: 1},
		}
	},
},
{input: `tag{.`, err: unexpectedTokenError{actual: Token{kind: tokenEnd, val: "."}}},
{input: `tag{x}.`, err: unexpectedTokenError{actual: Token{kind: tokenCloseCurly, val: "}"}}},
{input: `tag{x:}.`, err: unexpectedTokenError{actual: Token{kind: tokenCloseCurly, val: "}"}}},
{input: `tag{1:2}.`, err: unexpectedTokenError{actual: Token{kind: tokenInteger, val: "1"}}},
{input: `tag{x: ,}.`, err: unexpectedTokenError{actual: Token{kind: tokenComma, val: ","}}},
{input: `tag{x:1 y:2}.`, err: unexpectedTokenError{actual: Token{kind: tokenLetterDigit, val: "y"}}},
```

File: engine/dict.go [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+
+func (d *dict) Value(key Atom) (Term, bool) {
+	for k, v := range d.All() {
+		if k == key {
+			return v, true
+		}
+	}
+
+	return nil, false
+}
```
[the remark this answers is not in the document]
Comment (ccamel) [reply engine/dict.go:161]: Sure. That what I explained in the PR description. The use of a list of terms for Dict representation was chosen because it leverages Prolog’s pattern matching and unification capabilities through its compatibility with compound terms. However, I also suggest that there is room for improvement in this approach to enhance performance.

Please open a GitHub issue to address this matter. Thx.

File: engine/dict.go [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+		j += 2
+	}
+
+	return &dict{
+		compound: compound{
+			functor: atomDict,
+			args:    args,
+		},
+	}, nil
+}
```
[the remark this answers is not in the document]
Comment (ccamel) [reply engine/dict.go:349]: An invariant of `Dict` is that the keys in the compound arguments are sorted. This is done by the `Dict` constructor. Thus `All()` is guarantee to returns the key-value pairs in sorted order. That’s why I’m using a merge algorithm here, which combines two sorted lists into a single sorted list in linear time.

File: engine/parser.go [hunk 4]
```diff
@@ -506,6 +541,12 @@
 	case tokenFloatNumber:
 		return float(1, t.val)
 	case tokenVariable:
+		if t, _ := p.next(); t.kind == tokenOpenCurly {
+			p.backup()
+			p.backup()
+			return p.dict()
+		}
+		p.backup()
```
[the remark this answers is not in the document]
Comment (ccamel) [reply engine/parser.go:549]: Does it really deserve it? I mean, creating a function to avoid 4-5 repeated lines that don’t have a clearly defined business logic?

File: engine/dict.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ccamel) [reply engine/dict.go:161]: Oh, it seems you also opened issue #341 on `ichiban/prolog`. Please avoid doing that. You should close it.

**Comment by ccamel** [comment]
@coderabbitai pause

File: engine/dict.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ccamel) [reply engine/dict.go:161]: I can’t do that because I’m not the owner of the repository.

**Comment by triska** [comment]
To make this project useful for organizations and civil&nbsp;engineers that must&nbsp;adhere to international&nbsp;standards, I recommend keeping to the Prolog ISO&nbsp;standard. This will simplify procurement, teaching, and resolution of warranty&nbsp;disputes for your business, among other benefits such as alignment with WTO&nbsp;principles like nondiscrimination for companies and governments that adopt your&nbsp;solutions.

It is possible to add an infix `.` as a **conforming extension**, the following document explains how:

https://www.complang.tuwien.ac.at/ulrich/iso-prolog/SWI7_and_ISO

**Comment by ccamel** [comment]
@coderabbitai resolve

**Comment by ccamel** [comment]
@triska Oh, interesting! The project has already diverged a bit from the Prolog ISO standard because we had to adapt it for a blockchain environment. This introduced specific constraints, which led us to fork the original ichiban/prolog project and resulted in these deviations.

That said, I agree that staying as close to the standard as possible is the right approach. I’ll take a closer look at the document and aim to make an implementation that respects the standard. Thanks for sharing it!
